### PR TITLE
Immutable numeric index

### PR DESCRIPTION
**Pull request by IvanPleshkov**
Immutable state of numeric index.

The immutable state provides more efficient RAM usage. You can find the measurements below.
Like in Map Index, more efficient RAM usage was achieved because the immutable state doesn't support the addition of values. But deletion works as well.

### RAM difference measurement

RAM usage was measured using BFB (branch `skip-payload-indexing-option` with skipping field indices feature).

Measurement on dev with numeric index:
args: `bfb --float-payloads`
RAM: `262MB`

Measurement on dev without numeric index (but with payload):
args: `bfb --float-payloads --skip-field-indices`
RAM: `246MB`

Measurement on branch with numeric index:
args: `bfb --float-payloads`
RAM: `252MB`

Numeric index usage on dev: `262MB - 246MB = 16`
Numeric index usage on branch: `252MB- 246MB = 6`
It's almost 3 times better.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by generall — APPROVED** [review]

**Review by timvisee — APPROVED** [review]

File: lib/segment/src/index/field_index/numeric_index/mod.rs [hunk 1]
```diff
@@ -36,57 +37,88 @@ const HISTOGRAM_PRECISION: f64 = 0.01;
...
+            return std::cmp::Ordering::Equal;
+        }
+        if self.is_nan() {
+            return std::cmp::Ordering::Less;
+        }
+        if other.is_nan() {
+            return std::cmp::Ordering::Greater;
+        }
+        self.partial_cmp(other).unwrap()
+    }
```
[the remark this answers is not in the document]
Comment (IvanPleshkov) [reply lib/segment/src/index/field_index/numeric_index/mod.rs:80]: This function has to be guaranteed to have the same order as encoded `Vec<u8>` value

File: lib/segment/src/index/field_index/numeric_index/immutable_numeric_index.rs [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+impl<T> From<Point<T>> for NumericIndexKey<T> {
+    fn from(key: Point<T>) -> Self {
+        NumericIndexKey {
+            key: key.val,
+            idx: key.idx as PointOffsetType,
+            deleted: false,
+        }
+    }
+}
+
```
Comment (agourlay) [inline lib/segment/src/index/field_index/numeric_index/immutable_numeric_index.rs:85]: @IvanPleshkov  I am running Rust 1.73 locally and Clippy complains about this comparator.

```
error: incorrect implementation of `partial_cmp` on an `Ord` type
  --> lib/segment/src/index/field_index/numeric_index/immutable_numeric_index.rs:86:1
   |
86 | /  impl<T: PartialEq + PartialOrd + Encodable> PartialOrd for NumericIndexKey<T> {
87 | |      fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
   | | _______________________________________________________________________-
88 | ||         Some(match self.key.cmp_encoded(&other.key) {
89 | ||             std::cmp::Ordering::Equal => self.idx.cmp(&other.idx),
90 | ||             ord => ord,
91 | ||         })
92 | ||     }
   | ||_____- help: change this to: `{ Some(self.cmp(other)) }`
93 | |  }
   | |__^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#incorrect_partial_ord_impl_on_ord_type
   = note: `#[deny(clippy::incorrect_partial_ord_impl_on_ord_type)]` on by default
```

Comment (agourlay) [reply lib/segment/src/index/field_index/numeric_index/immutable_numeric_index.rs:85]: FYI applying the proposed fix is not great

```
thread 'index::field_index::numeric_index::immutable_numeric_index::tests::test_numeric_index_key_set' has overflowed its stack
fatal runtime error: stack overflow
```

Comment (timvisee) [reply lib/segment/src/index/field_index/numeric_index/immutable_numeric_index.rs:85]: This fix is implemented in <https://github.com/qdrant/qdrant/pull/2767>.

**Pull request merged**